### PR TITLE
VIP28: Generalize type methods

### PR DESCRIPTION
**Pull request by Dridi**
This is not a complete implementation of what we've discussed regarding VIP28, but only a first step.

With this patch set VMOD objects now have a dedicated type, and any type may have methods. This is handled via the symbol table, consider the following snippet:

```vcl
new fb1 = directors.fallback();
new fb2 = directors.fallback();
```

Currently you could get the following entries in the symbol table:

```
object    VOID      41 41 directors.fallback
instance  INSTANCE  41 41 fb1
func      VOID      40 41 fb1.add_backend
func      BACKEND   40 41 fb1.backend
func      VOID      40 41 fb1.remove_backend
instance  INSTANCE  41 41 fb2
func      VOID      40 41 fb2.add_backend
func      BACKEND   40 41 fb2.backend
func      VOID      40 41 fb2.remove_backend

```

With this change it would be reduced to the following:

```
object    directors.fallback  41 41 directors.fallback
method    VOID                40 41 directors.fallback::add_backend
method    BACKEND             40 41 directors.fallback::backend
method    VOID                40 41 directors.fallback::remove_backend
instance  directors.fallback  41 41 fb1
instance  directors.fallback  41 41 fb2
```

With the current methods available to the native VCL types, the following entries are now always present in the symbol table:

```
method    BYTES      40 41 STEVEDORE::free_space
method    BOOL       40 41 STEVEDORE::happy
method    BYTES      40 41 STEVEDORE::used_space
method    STRING     40 41 STRINGS::lower
method    STRING     40 41 STRINGS::upper
```

Other than shaping the internals for future VIP28 material, this patch set doesn't change any VCL behavior.

### CODE REVIEW

**Comment by Dridi** [comment]
Oh, and it doesn't conflict with #3116 or #3123.

**Comment by rezan** [comment]
Do we have a proposal of how types will actually look in VCL?

So I did finally wrap my head around VIP28. The one thing I didn't like is that is blended objects and variables. Basically is reuses objects as variable name holders. So is there a new idea to take VIP28's place?

I think if we are going to be successful here, we need to start with how we want the syntax to look and behave and then we implement it. If we start with patching vcc to extend it into variables, then I feel we will end up with a not so great final product.

To be (brutally) honest, vcc as it stands today is not the best platform for building a strong language. Its not the whole VCL to C idea, i think thats great. Rather, its just the way vcc goes about it and organizes itself. I almost sometimes feel like rewriting the entire thing. So if we find its not possible to have variables the way we want then using the current vcc, then we know why.

**Comment by Dridi** [comment]
> Do we have a proposal of how types will actually look in VCL?

I suppose VIP 28 gives a broad view. I was able to squeeze a bit more details about the general idea from @bsdphk on IRC.

> So I did finally wrap my head around VIP28. The one thing I didn't like is that it blended objects and variables. Basically is reuses objects as variable name holders. So is there a new idea to take VIP28's place?

I'm not sure what you are referring to. The VIP28 @nigoroll submitted on our behalf was discussed and vetoed last bugwash. @bsdphk suggested a different approach.

> I think if we are going to be successful here, we need to start with how we want the syntax to look and behave and then we implement it. If we start with patching vcc to extend it into variables, then I feel we will end up with a not so great final product.

On the other hand unless you patch libvcc nothing is going to happen. The pull request description makes it very clear that this is only about changing the internal representation of related VCL elements, not behavior. This builds foundations for the things discussed around VIP28 and erases special cases and most notably differences between VMOD `$Method`s and type methods.

> To be (brutally) honest, vcc as it stands today is not the best platform for building a strong language. Its not the whole VCL to C idea, I think thats great. Rather, its just the way vcc goes about it and organizes itself. I almost sometimes feel like rewriting the entire thing.

I will be equally brutal and also not back my claim with any form of evidence: from-scratch projects rewrites tend to be utter failures. With this patch set at least I made enough progress to reach a point where `vcc_expr4()` is the only place where we evaluate method calls in expressions.

> So if we find its not possible to have variables the way we want when using the current vcc, then we know why.

I don't think this ticket is the right place for this discussion.

Incidentally I added some coverage in dcf4aaacb273f4d638cdc7e4b2675b9b10597d0e when I realized this syntax error was not checked by our test suite.

**Comment by rezan** [comment]
Ok, I guess I wasn't sure on the status of VIP28.

Why not just implement types as plain `long`, `double`, and `char*` C variables? Declare them in the sub, auto detect the actual type on initialization, and then start using them in expressions as simple typed symbols. Seems almost too simple...?

**Comment by Dridi** [comment]
Again, this is completely irrelevant for this pull request, maybe #3141 is a better venue for such a discussion.

I will probably remove d40bf658def05626d3f18b8a99bc8ca1a2333905 from the patch set. I don't remember why it made sense to rename `struct type` in the first place.

File: lib/libvcc/vcc_expr.c [hunk 1]
```diff
@@ -848,14 +823,14 @@ vcc_expr4(struct vcc *tl, struct expr **e, vcc_type_t fmt)
...
-		for(vm = vcc_methods; vm->type_from != NULL; vm++) {
-
-			if (vm->type_from == (*e)->fmt &&
-			    vcc_IdIs(tl->t, vm->method))
+		vm = (*e)->fmt->methods;
+		while (vm != NULL && vm->type != NULL) {
+			if (vcc_IdIs(tl->t, vm->name))
 				break;
+			vm++;
 		}
```
[the remark this answers is not in the document]
Comment (Dridi) [reply lib/libvcc/vcc_expr.c]: Will do, but in `vcc_types.c` where this code ultimately ends up.

File: lib/libvcc/vcc_vmod.c [hunk 2]
```diff
@@ -452,28 +538,4 @@ vcc_Act_New(struct vcc *tl, struct token *t, struct symbol *sym)
...
-		VSB_clear(buf);
-		VSB_printf(buf, "%s.%s", sy1->name, vf->value);
-		AZ(VSB_finish(buf));
-		sy3 = VCC_MkSym(tl, VSB_data(buf), SYM_FUNC, VCL_LOW, VCL_HIGH);
-		AN(sy3);
-		func_sym(sy3, sy2->vmod_name, VTAILQ_NEXT(vf, list));
-		sy3->extra = p;
-		vv = VTAILQ_NEXT(vv, list);
-	}
-	VSB_destroy(&buf);
```
[the remark this answers is not in the document]
Comment (Dridi) [reply lib/libvcc/vcc_vmod.c]: The last `vv` assignment is then read in the loop condition, unless I'm missing something else?

**Comment by Dridi** [comment]
> I really dont like the "::" stunt, there must be a better way to do that...

I'm open to suggestions. The goal was to have the ability to look up methods of a given expression, to ultimately be able to leverage type methods to implement basic operators (as described in https://github.com/varnishcache/varnish-cache/issues/3141#issuecomment-558203838).

The problem I ran into with the dot notation is that `int` is currently a valid symbol name so `new int = vmod.constructor(...);` is allowed, so I can't put `INT.add(expr1, expr2)` in the symbol table as `INT.add`.

Or if I do that, I need to introduce a new 4.2 or 5.0 VCL syntax where this becomes a restriction, which wasn't the goal of this pull request. I wanted to generalize methods evaluation to allow us to start working on generalizing operators using type methods.

So I looked at python and found nothing and fell back to the C++ double-colon notation.

**Comment by Dridi** [comment]
I took care of the reviews item so far and pushed an additional cosmetic patch: c5d0c979835e0a7d0e19640d10b955ad1409d65a.

**Comment by bsdphk** [comment]
I think it would be ok-ish to ban typenames as object names in vcl4.1.

**Comment by nigoroll** [comment]
Re @rezan 

First of all, the subject of @bsdphk s criticism, at least how I understand it, was the `$Cast` model, which was a deliberate limitation of the more general idea of type methods. This is where VIP28 comes from and where it seems to be heading.

> The one thing I didn't like is that it blended objects and variables. Basically is reuses objects as variable name holders.

Absolutely. If we want to become "more OO", using objects as variables makes a lot of sense, because variables _are_ objects (the most trivial ones). If you say you "don't like" it, why don't you? Why do you think variables would need to be a different animal?

> To be (brutally) honest, vcc as it stands today is not the best platform for building a strong language.

I do not think there is any limitation in the varnish-cache infrastructure which would keep you from plugging whatever compiler/interpreter in as a .so file and run that. So if you want something other than VCL, you can have that, just write it.

But I must say there are some things about VCL which I find extremely valuable with respect to the purpose it serves, and I do think these aspects justify for an own domain specific language with its own limitations. This list is sure incomplete, but just to mention what comes to mind immediately:
* no loops (except for retry/restart) are an important safeguard against self-dos
* implementations tailored to the lifetime model (basically the vcl vs. workspace concept) allow for efficient memory management

So at this point I think that VCL is a language which is extremely well suited for it's purpose (far better than anything else I have seen in today's CDNs and far betten than Webassembly which, as far as I understand, some are pursuing to replace VCL).

**Comment by nigoroll** [comment]
bugwash: please invest some more work

**Comment by Dridi** [comment]
bugwash: also make VCL types reserved keywords, so a symbol may not be called `int` or `string`.

**Comment by Dridi** [comment]
We kinda painted ourselves into a corner here, we already have at least one first-party symbol named after a type: `vmod_blob`. (and I must say, one that is very hard to follow!)

Once you `import blob` in your VCL you get the following entry in the symbol table:

```
vmod      VOID       41 41 blob*
```

We could maybe get away with it for VMODs and pretend that wildcard symbols are not the same as their plain counterpart:


```
reserved  VOID       4X 4X blob
vmod      VOID       4X 4X blob*
```

@bsdphk, would you find this acceptable? I'm not entirely sure what this change entails, but it could look something like below. Beware, this is actually accurate so make sure not to read this right after a meal to avoid a blob soup indigestion:

```
kind     | type      | name             | parent
---------+-----------+------------------+----------
reserved | VOID      | BLOB             | <none>
vmod     | VOID      | blob*            | BLOB
object   | blob.blob | blob.blob        | blob*
method   | BLOB      | blob.blob.get    | blob.blob
method   | STRING    | blob.blob.encode | blob.blob
```

Symbol names are case insensitive, so could end up with either case for the BLOB symbol itself. I used the uppercase to make it stand out a bit because I was having a hard time following my own trail of thought (and the fancy ascii art table helped a bit too).

I will need to check what the consequences are for other types of wildcard symbols, and we'd need to tolerate that a wildcard might not have a parent with the same name. I'll do that research later, I need an opinion first.

**Comment by bsdphk** [comment]
Maybe the right thing to do is to flesh out the symboltables for vmods on import, rather than rely on the wildcard symbols ?

**Comment by Dridi** [comment]
I very much agree. This is what this pull request does with object methods: the first time we run into a constructor we register all methods at once. However doing that I had to monkey-patch the symbol table code a bit because the symbols I was actively registering were also tripping on themselves at creation time for being under a wildcard symbol...

So virtually `blob.*` symbols from `vmod_blob` would be considered children of the `blob` reserved type symbol, even though they aren't related. I can change tack in that direction, I will revisit this later if we agree on the destination.

File: lib/libvcc/vcc_vmod.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dridi) [reply lib/libvcc/vcc_vmod.c]: I see it now, temporarily fixed in cc6b86383a9be1d7402dc9898710cd0b4d7f2b16.

**Comment by Dridi** [comment]
I hope it won't bother anybody but I'm locking this thread. This pull request served its purpose and we now have a better idea of where we want to go, and I will do that as a mix of direct push to master and smaller incremental pull requests. I'm keeping this one open until its full scope has landed in master.

**Comment by Dridi** [comment]
On second thought, we can't really do what I suggested. If we start adding methods to a type, and allow VMODs to be named after types, like vmod_blob, then we run into the risk of symbol conflicts (besides the base symbol) where type methods could conflict with VMOD functions or object constructors.

We could rename vmod_blob to something like vmod_blobs, or we could retire it as suggested in https://github.com/varnishcache/varnish-cache/pull/3123#issuecomment-563328719. There is one thing from vmod_blob that we can't replace with type methods: `$Object blob(ENUM, STRANDS)`. I suggest we move that to vmod_std instead (`new foo = std.blob(...);`).

But before we get there, type methods need to grow the ability to take named argument with potential default values? The prep work for this to become actionable keeps piling up but I'm still up to the task :)

@slimhazard @nigoroll @bsdphk, thoughts?

PS. locking the thread was pointless, it doesn't apply to organization members

**Comment by nigoroll** [comment]
@Dridi can you please help me understand the problem? Can't we just make classes be types, in which case class methods _are_ type methods?

**Comment by Dridi** [comment]
> @Dridi can you please help me understand the problem?

The way I implemented this patch series with `${type}::${method}` symbols did not conflict because there was no need for a `${type}` symbol. So the `SYM_VMOD` symbol for vmod_blob (called `blob`) wouldn't conflict with a counterpart `SYM_RESERVED` symbol for the blob type (also called `blob`).

If we put `${type}.${method}` symbols as requested by @bsdphk we then need that `${type}` symbol to exist because the symbol table is hierarchical as shown in the table from https://github.com/varnishcache/varnish-cache/pull/3147#issuecomment-561064504 (the parent relationship). 

> Can't we just make classes be types, in which case class methods are type methods?

If you look at the pull request description (https://github.com/varnishcache/varnish-cache/pull/3147#issue-347168277) that's what I did:

```
object    directors.fallback  41 41 directors.fallback
method    VOID                40 41 directors.fallback::add_backend
method    BACKEND             40 41 directors.fallback::backend
method    VOID                40 41 directors.fallback::remove_backend
instance  directors.fallback  41 41 fb1
instance  directors.fallback  41 41 fb2
```

The constructor's symbol (`SYM_OBJECT`) has a type for the return value, and this type is created on demand, when a `new` statement is encountered. And `$Method`s are created as `SYM_METHOD` symbols.

Then instances inherit the type returned by the constructor, and that's how they get access to the methods.

**Comment by slimhazard** [comment]
Some mixed and contradictory thoughts ...

Besides the language-technical issue of symbol conflicts, I think it would make sense if the VIP28-style properties and methods for data types end up replacing VMOD blob. The VMOD is all about a data type, says so right in the name. If language support is introduced for what it does for the data type, then we have two ways of doing the same thing, and we only need one.

However, I was hoping that a demise of VMOD blob could come softly, to ease the transition effort -- say, announce that the VMOD is deprecated, but leave it in the installation for a couple of release cycles. A new version that abruptly removes a VMOD would mean that some of us have to rewrite a lot of VCL in order to upgrade. Not as hard as the transition from 3.x -> 4.x (let's hope that record never gets broken), but still significant. Remember that the changes necessary for 3.x -> 4.x meant that many more deployments remained "stuck" in 3.x for longer than we would have liked.

BLOBs probably see very different usage in different projects. I suspect that many projects never touch BLOBs or the VMOD, and they wonder what that esoteric weirdness is good for. But @nigoroll and I work on projects in which they're used for digests, HMACs and the like -- a bit of advanced VCL, to be sure, but not outlandish for an HTTP application. They're important features that make the Varnish deployment valuable.

So that's my contradictory answer. I can see the technical problem, and why a move away from the VMOD makes sense. But although I frankly don't know how, I would hope there's a way to leave the VMOD in as a deprecated feature, rather than drop it all at once.

**Comment by Dridi** [comment]
If we want to ease the transition, we can have it renamed to `vmod_blobs` and deprecated once type methods grow more capabilities. Remember that today the type methods are very, very, very limited so renaming first to open the way towards type methods generalization would likely be more sensible.

**Comment by slimhazard** [comment]
Well, not ideal (feels a little hackish), but if there's no other way, adding an 's' is not the hardest thing to do.

**Comment by nigoroll** [comment]
@Dridi I hear what you are saying about the implementation you started, but, again, I do not understand why we can't just retire the _counterpart `SYM_RESERVED` symbol for the blob type (also called blob)_ and add the type methods to access blobs.
I guess we would need to take some more steps at once (for instance, vmods would need to learn dependencies, `BLOB` arguments would change into `blob.blob` arguments), but I still do not see why, conceptually, types and their classes would be different things.

**Comment by Dridi** [comment]
> @Dridi I hear what you are saying about the implementation you started, but, again, I do not understand why we can't just retire the counterpart `SYM_RESERVED` symbol for the blob type (also called blob) and add the type methods to access blobs.

This is getting confusing. There is currently no `SYM_RESERVED` symbol called blob, and certainly not in my implementation.

@bsdphk was not happy with my `::` workaround, so I experimented with the plain `.` ancestry separator. If we do that, then we have a `SYM_RESERVED` symbol for each type including `blob` and trying to `import blob;` fails. The rest of the test suite passed with flying colors.

This, is the corner we painted ourselves into.

Interestingly enough, putting vmod_blob aside, even with the `.` separator I ran into another conflict. We have symbols in the table that really are keywords, for example `acl` and `backend`. They both are symbols used to detect VCL top-level "actions" (declaring a backend or an acl) but they also become type symbol (that may have method children). Another conflict, but this one I managed to work around because I'm evil.

This, is the other corner we painted ourselves into, which I failed to mention since I was focused on my inability to reconcile this model for vmod_blob.

> I guess we would need to take some more steps at once (for instance, vmods would need to learn dependencies, BLOB arguments would change into blob.blob arguments),

I'm even more confused. What do you mean about dependencies? Because of the parent/child relationship of symbols `my_vmod.my_func` (`SYM_FUNC`) is already dependent on `my_vmod` (`SYM_VMOD`).

> but I still do not see why, conceptually, types and their classes would be different things.

Let's recap:

- Today in trunk we don't have anything of the sort and types don't appear in the symbol table, so nothing prevents us from using a type name for some other purpose, like we do with vmod_blob.
- This pull request doesn't give symbols to native types, but `SYM_OBJECT` and `SYM_INSTANCE` grew dedicated types instead of respectively `VOID` and `INSTANCE` (which conceptually does not differentiate types and their classes).
- My experiment to drop `::` added symbols for native types in addition to VMOD objects and conceptually there is also no difference. We painted ourselves into 2 corners, and my hoverboard didn't have enough fuel to escape both during this experiment.

Are you also getting increasingly more confused as the discussion progresses? :p

**Comment by Dridi** [comment]
#3158
